### DISCUSSION TITLE
Annotate optional list of items as query params

### DISCUSSION THREAD

**Discussion by developer992** [pydantic/pydantic#9569] [Question] [answered]
Hello,

i am having some hard times trying to figure this out:

Here is what I need:

an endpoint that is able to look by one ID or several
for instance:

Search by one id: GET on localhost:8000/endpoint/?id=1

Search by several ids:
Option 1:
GET on localhost:8000/endpoint/?ids=1,2,3

Option 2:
GET on localhost:8000/endpoint/?ids=1&ids=2&ids=3

What I tried:
Attempt 1:
```
class ItemFilterDTO (BaseModel):
    ids: Annotated[List[str], Query()] = Field(Query(default=[]))
```

This works as Option 2 and also nicely displays on Swagger docs **BUT** it complains with "cannot specify 'Annotated' and value Fields together for ids". Seems to work despite the warning in my IDE though ( Intelli J )

Swagger shows this like so:

![image](https://github.com/pydantic/pydantic/assets/149699593/9f83d24a-2cda-4df1-87b5-ff41ecce6654)

Attempt 2:
```
class ItemFilterDTO (BaseModel):
    ids: Annotated[List[str] | None, Field(Query(None))]
```

This works as Option 1 BUT if you provide several ids then the value of the field is ["1,2,3"] and not ["1", "2", "3"]

Swagger:
![image](https://github.com/pydantic/pydantic/assets/149699593/f58e5866-4dd4-47bf-9806-a705a48ca21d)

I suppose i could split the string my self but i feel like this should work out of the box ... i am probably just not using it correctly, right?

the FastAPI endpoint is quite simple:

```
@router.get("/item/list")
async def list_items(
    filters: ItemFilterDTO = Depends(),
) -> PaginatedItemsDTO:
    ...
```

These also don't work as expected.
`ids: list[str] | None = Field(Query(None))` - this also does Option 1 with ["1,2,3"]

Thanks!

**Comment by developer992** [comment]
Ok some progress:

this seems to work

```
@router.get("/item/list")
async def list_items(
    ids: list[str] | None = Query(None, title="List of IDs", description="fetch list"),
) -> PaginatedItemsDTO:

    print(ids)
```
`curl "http://localhost:8000/ref/material/list"` prints None
`curl "http://localhost:8000/ref/material/list?ids=1"` prints ids=["1"]
`curl "http://localhost:8000/ref/material/list?ids=1&ids=2"` prints ids=["1", "2"]

But swagger does not like this, renders like this:
![image](https://github.com/pydantic/pydantic/assets/149699593/e39a5d83-24ba-4392-bcd9-29930bd7d482)

Cannot test with more than one ID, but i can live without that.

but not this
```
class Foo(BaseModel):
    ids: list[str] | None = Query(None, title="List of IDs", description="fetch list"),

@router.get("/item/list")
async def list_items(
    filters: Foo = Depends(),
) -> PaginatedItemsDTO:
```
Not sure if it's a problem with fastapi or pydantic though.

**Comment by chyok** [comment, marked as answer]
Hi @developer992 ,

For Attempt 1, I think the implementation should be correct, and I don't see any warnings in my PyCharm. You could try upgrading to the latest version of PyCharm.

For Attempt 2, if you want to accept a query parameter like "1,2,3", you could try something like this:
```python
async def query_parameters(ids: str | None = None):
    return ids.split(",")


@router.get("/item/list")
async def list_items(
        filters: Annotated[List[str], Depends(query_parameters)],
):
    print(filters)  # ['1', '2']
    ...
```

<img width="347" alt="image" src="https://github.com/pydantic/pydantic/assets/32629225/b6474ff0-1be4-473c-9133-47d316189936">

**Comment by developer992** [comment]
Great thanks! I will definitely try updating and/or PyCharm. This split string as dependency is very elegant too, i like it, much appreciated!
